source/common/http/conn_manager_impl.cc

```diff
@@ -423,23 +423,36 @@ void ConnectionManagerImpl::onEvent(Network::ConnectionEvent event) {
 
   if (event == Network::ConnectionEvent::RemoteClose ||
       event == Network::ConnectionEvent::LocalClose) {
-    if (connection_idle_timer_) {
-      connection_idle_timer_->disableTimer();
-      connection_idle_timer_.reset();
-    }
+    // TODO(mattklein123): It is technically possible that something outside of the filter causes
+    // a local connection close, so we still guard against that here. A better solution would be to
+    // have some type of "pre-close" callback that we could hook for cleanup that would get called
+    // regardless of where local close is invoked from. Note that this will cause this method to
+    // get called twice in the common local close cases, but the method protects against that.
+    doConnectionClose(absl::nullopt);
```

Comment:
so if we call doConnectionClose with absl::nullopt, we're going to be missing stream stats on cleanup.  Do you think this should never happen (at which point we can assert that streams_ is empty), or can this actually happen (and we need a separate value for "should we close at the end of clean up?")

Reply:
We won't actually miss stats, they would just potentially be shared to remote instead of local. This occurred to me, but I decided to ignore it, but of course you have a very sharp eye. :) My personal opinion is that it's not worth handling this case right now given I'm trying to be paranoid about this path. I thought about an ASSERT also, but that seemed a bit risky. Maybe just more comments? WDYT?